### PR TITLE
Allow get_type_hints to work with dataclasses where __init__ is created automatically and using delayed annotations

### PR DESCRIPTION
**Pull request by georgeharker**
as per https://bugs.python.org/issue34776

there can be issues when doing something like:

```
from __future__ import annotations
from typing import Optional, get_type_hints

@dataclass
class Foo:
   a: Optional[int]


get_type_hints(Foo.__init__)

```

will give

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/typing.py", line 1001, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/usr/lib/python3.7/typing.py", line 260, in _eval_type
    return t._evaluate(globalns, localns)
  File "/usr/lib/python3.7/typing.py", line 464, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'Optional' is not defined
```

Which causes such classes not to be able to read in by jsons.

A workaround is to ensure that get_type_hints has the module dict of the class available to it.

### CODE REVIEW

**Comment by georgeharker** [comment]
I've made the tests appropriately version-specific, ensured it does tickle the issue.  I can't figure out for the life of me why the coverage drops, it should be hitting the new lines (unless the coverage isn't run under 3.7)

File: tests/test_specific_versions.py [hunk 1]
```diff
@@ -51,6 +51,25 @@ def test_simple_dump_and_load_dataclass_with_future_import(self):
 
         self.assertDictEqual(expected, dumped)
 
+    @only_version_3(7, and_above=False)
+    def test_postponed_annoation_dataclass(self):
+        from postposned_dataclass import A
```
Comment (meshy) [inline tests/test_specific_versions.py:56]: There's a typo in the module name that's leaked into the code here.

```suggestion
        from postponed_dataclass import A
```

**Comment by georgeharker** [comment]
Oops. I’ll take a look at fixing that. Autocomplete is great until it
autocompletes typos.

George

**Comment by georgeharker** [comment]
Updated now I figured out how to properly exclude the error in 3.5

**Review by ramonhagenaars — APPROVED** [review]
Thanks @georgeharker for this pull request. I'd like to add it to the `1.4.0` release.

**Pull request merged**